### PR TITLE
feat: deployment-types

### PR DESCRIPTION
**Pull request by linbuxiao**
- [x] deployment-types

The ice-cream story.

### CODE REVIEW

**Comment by linbuxiao** [comment]
I noticed that the `package-lock` had been modified. The first thing I want to make sure is `lock` should be uploaded?

**Comment by WingLim** [comment]
> I noticed that the `package-lock` had been modified. The first thing I want to make sure is `lock` should be uploaded?

I take a look the lock file, you should not update it.
Because it roll back to version 1 again, it cause by the npm version as I did last time.

**Comment by KnorpelSenf** [comment]
Maybe have a look at https://github.com/nvm-sh/nvm. It's the only way I know that makes handing different versions of Node and npm feasible. It gets super simple to select the right version of npm such that this doesn't happen.

File: site/docs/zh/guide/deployment-types.md [hunk 1]
```diff
@@ -1,3 +1,387 @@
...
+_想象一下，你在你喜爱的冰淇淋店买了一勺冰淇淋，
+你走向服务员点了一份最喜欢的冰淇淋。
+但不走运的是，这种冰淇淋已经售罄了。_
+
+_第二天，你又想吃那个美味的冰淇淋了，所以你又回到同一个地方，点了同样的冰淇淋。
+好消息！
+他们在昨晚进货，现在你可以吃美味的冰淇淋了！
+这真的是非常美味的冰淇淋。_
+
+**长轮询**意味着 grammY 主动的给 Telegram 服务器发送了一个请求，来询问是否有新的更新（或许是消息什么的）。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:39]:
```suggestion
**长轮询**意味着 grammY 主动的给 Telegram 服务器发送了一个请求，来询问是否有新的 update（或许是消息什么的）。
```

File: site/docs/zh/guide/deployment-types.md [hunk 2]
```diff
@@ -1,3 +1,387 @@
...
+如果没有消息，Telegram 将返回一个空列表。
+这个的意思是从你上一次请求到这一次请求之间并没有新的消息发送给你的 bot 。
+
+当 grammY 发送请求给 Telegram 服务器时，有新的消息发送给了你的 bot，Telegram 将把这些更新以最多 100 条的更新以数组的形式发送给你。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:43]:
```suggestion
当 grammY 发送请求给 Telegram 服务器时，有新的消息发送给了你的 bot，Telegram 将以最多 100 条的 update 对象的数组发送给你。
```

File: site/docs/zh/guide/deployment-types.md [hunk 3]
```diff
@@ -1,3 +1,387 @@
...
+使用 `bot.start ()`，一切都会正常工作，不需要进一步配置。
+在负载下，您可以完全控制需要处理消息的个数。
+
+长轮询可以在这些地方良好运行：
+
+- 可以在你的本地，
+- 所有的 VPS，或是
+- 在后端运行的 bot 示例上（即一个全天候运行的机器）都可以良好运行。
+
+**Webhooks 比长轮询的主要优势在于它们更”便宜“。**
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:149]:
```suggestion
**Webhooks 比长轮询的主要优势在于它们更“便宜”。**
```

File: site/docs/zh/guide/deployment-types.md [hunk 4]
```diff
@@ -1,3 +1,387 @@
...
+## 如何使用长轮询
+
+使用
+
+```ts
+bot.start();
+```
+
+你使用一个非常简单的长轮询运行你的 bot。
+它按顺序处理所有更新。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:180]:
```suggestion
它按顺序处理所有 update。
```

File: site/docs/zh/guide/deployment-types.md [hunk 5]
```diff
@@ -1,3 +1,387 @@
...
+```
+
+</CodeGroupItem>
+</CodeGroup>
+
+如果你考虑在 VPS 上使用 webhooks 运行你的 bot，请确保你阅读了 Telegram 团队写的 [Marvin's Marvellous Guide to All Things Webhook](https://core.telegram.org/bots/webhooks)。
+
+### Webhook Reply
+
+当收到一个 webhook 请求时，您的 bot 可以在响应中调用一个方法。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:238]:
```suggestion
当收到一个 webhook 请求时，你的 bot 可以在响应中调用一个方法。
```

File: site/docs/zh/guide/deployment-types.md [hunk 6]
```diff
@@ -1,3 +1,387 @@
...
+这样做的一个好处是，每次更新都可以节省一个 HTTP 请求的 bot。然而，使用这种方法也有一些缺点:
+
+1. 您将无法处理各个 API 调用的潜在错误。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:241]:
```suggestion
1. 你将无法处理各个 API 调用的潜在错误。
```

File: site/docs/zh/guide/deployment-types.md [hunk 7]
```diff
@@ -1,3 +1,387 @@
...
+   这包括速率限制错误，因此实际上不能保证您的请求有任何效果。
+2. 更重要的是，您也无法访问响应对象。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:243]:
```suggestion
2. 更重要的是，你也无法访问响应对象。
```

File: site/docs/zh/guide/deployment-types.md [hunk 8]
```diff
@@ -1,3 +1,387 @@
...
+   例如，调用 `sendMessage` 将不能使您访问您发送的消息。
+3. 此外，不可能取消请求。
+   `abotsignal` 将被忽略。
+4. 还要注意的是，grammY 中的类型并不反映所执行的 webhook 回调的结果！
+   例如，它们表明您总是接收到一个响应对象，但确保在使用这个次要的性能优化时，您将自己责任，努力让他不会出错。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:248]:
```suggestion
   例如，它们表明你总是接收到一个响应对象，但确保在使用这个次要的性能优化时，努力让它不会出错是你自己的责任。
```

File: site/docs/zh/guide/deployment-types.md [hunk 9]
```diff
@@ -1,3 +1,387 @@
...
+### 及时结束 Webhook 请求
+
+> 如果所有中间件都很快完成，也就是在几秒钟内完成，则可以忽略本页面的其余部分。
+> 本节主要为那些希望进行文件传输以响应消息或其他需要更多时间的操作的用户服务。
+
+当 Telegram 服务从一个聊天发送更新到你的 bot 服务器时，它会等待你结束请求，然后再发送属于该聊天的下一个更新。
+换句话说，Telegram 将按顺序发送来自同一聊天的更新，来自不同聊天的更新将同时发送。
+(这条消息的来源是[这里](https://github.com/tdlib/telegram-bot-api/issues/75#issuecomment-755436496).)
+
+Telegram 试图确保你的 bot 收到所有更新。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:288]:
```suggestion
Telegram 试图确保你的 bot 收到所有 update。
```

File: site/docs/zh/guide/deployment-types.md [hunk 10]
```diff
@@ -1,3 +1,387 @@
...
+|            | <--- 好的，执行 sendChatAction ---  |           |
+|____________|                                   |___________|
+```
+
+### 及时结束 Webhook 请求
+
+> 如果所有中间件都很快完成，也就是在几秒钟内完成，则可以忽略本页面的其余部分。
+> 本节主要为那些希望进行文件传输以响应消息或其他需要更多时间的操作的用户服务。
+
+当 Telegram 服务从一个聊天发送更新到你的 bot 服务器时，它会等待你结束请求，然后再发送属于该聊天的下一个更新。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:284]:
```suggestion
当 Telegram 服务从一个聊天发送 update 到你的 bot 服务器时，它会等待你结束请求，然后再发送属于该聊天的下一个 update。
```

File: site/docs/zh/guide/deployment-types.md [hunk 11]
```diff
@@ -1,3 +1,387 @@
...
+换句话说，Telegram 将按顺序发送来自同一聊天的更新，来自不同聊天的更新将同时发送。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:285]:
```suggestion
换句话说，Telegram 将按顺序发送来自同一聊天的 update，来自不同聊天的 update 将同时发送。
```

File: site/docs/zh/guide/deployment-types.md [hunk 12]
```diff
@@ -1,3 +1,387 @@
...
+(这条消息的来源是[这里](https://github.com/tdlib/telegram-bot-api/issues/75#issuecomment-755436496).)
+
+Telegram 试图确保你的 bot 收到所有更新。
+这意味着，如果某个聊天的更新传递失败，后续的更新将被排队，直到第一次更新成功。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:289]:
```suggestion
这意味着，如果某个聊天的 update 传递失败，后续的 update 将被排队，直到第一次 update 成功。
```

File: site/docs/zh/guide/deployment-types.md [hunk 13]
```diff
@@ -1,3 +1,387 @@
...
+
+#### 为什么不停止一个 Webhook 请求是危险的
+
+对于每次发送到你的 webhook 端点的更新，Telegram 都有一个超时时间。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:293]:
```suggestion
对于每次发送到你的 webhook 端点的 update，Telegram 都有一个超时时间。
```

File: site/docs/zh/guide/deployment-types.md [hunk 14]
```diff
@@ -1,3 +1,387 @@
...
+如果你不尽快结束一个 webhook 请求，Telegram 将假设它没有发送，于是重新发送了更新。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:294]:
```suggestion
如果你不尽快结束一个 webhook 请求，Telegram 将假设它没有发送，于是重新发送了 update。
```

File: site/docs/zh/guide/deployment-types.md [hunk 15]
```diff
@@ -1,3 +1,387 @@
...
+因此，你的 bot 可能会意外地多次处理同一个更新。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:295]:
```suggestion
因此，你的 bot 可能会意外地多次处理同一个 update。
```

File: site/docs/zh/guide/deployment-types.md [hunk 16]
```diff
@@ -1,3 +1,387 @@
...
+这意味着它将多次执行所有更新处理，包括发送任何响应消息。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:296]:
```suggestion
这意味着它将多次执行所有 update 处理，包括发送任何响应消息。
```

File: site/docs/zh/guide/deployment-types.md [hunk 17]
```diff
@@ -1,3 +1,387 @@
...
+|            |        *bot 处理了第三次*      ...  |           |
+|____________|                              ...  |___________|
+```
+
+这就是为什么 grammY 在 `webhookCallback` 中有着更短的超时时间(默认值: 10 秒) 。
+如果您的中间件在此之前完成，`webhookCallback` 函数将自动响应 webhook。
+那样的话，一切都很好。
+但是，如果你的中间件在 grammY 超时之前没有完成，`webhookCallback` 会抛出一个错误。
+这意味着您可以处理 web 框架中的错误。
+如果你没有处理那个错误，Telegram 将会再次发送同样的更新---- 但至少你现在有了错误日志，告诉你有什么地方出错了。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:320]:
```suggestion
如果你没有处理那个错误，Telegram 将会再次发送同样的 update ---- 但至少你现在有了错误日志，告诉你有什么地方出错了。
```

File: site/docs/zh/guide/deployment-types.md [hunk 18]
```diff
@@ -1,3 +1,387 @@
...
+|            |        *bot 处理了第二次*      ..   |           |
+|            |                              ..   |           |
+|            | ---      你好！！！！       --->    |           |
+|            |                              ...  |           |
+|            |        *bot 处理了第三次*      ...  |           |
+|____________|                              ...  |___________|
+```
+
+这就是为什么 grammY 在 `webhookCallback` 中有着更短的超时时间(默认值: 10 秒) 。
+如果您的中间件在此之前完成，`webhookCallback` 函数将自动响应 webhook。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:316]:
```suggestion
如果你的中间件在此之前完成，`webhookCallback` 函数将自动响应 webhook。
```

File: site/docs/zh/guide/deployment-types.md [hunk 19]
```diff
@@ -1,3 +1,387 @@
...
+那样的话，一切都很好。
+但是，如果你的中间件在 grammY 超时之前没有完成，`webhookCallback` 会抛出一个错误。
+这意味着您可以处理 web 框架中的错误。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:319]:
```suggestion
这意味着你可以处理 web 框架中的错误。
```

File: site/docs/zh/guide/deployment-types.md [hunk 20]
```diff
@@ -1,3 +1,387 @@
...
+如果你没有处理那个错误，Telegram 将会再次发送同样的更新---- 但至少你现在有了错误日志，告诉你有什么地方出错了。
+
+一旦 Telegram 第二次向你的 bot 发送更新，你处理它的速度不可能比第一次快。
+因此，它可能会再次超时，而 Telegram 将再次发送更新。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:323]:
```suggestion
因此，它可能会再次超时，而 Telegram 将再次发送 update。
```

File: site/docs/zh/guide/deployment-types.md [hunk 21]
```diff
@@ -1,3 +1,387 @@
...
+因此，你的 bot 不仅会看到两次更新，而是几十次，直到 Telegram 停止重新尝试。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:324]:
```suggestion
因此，你的 bot 不仅会看到两次 update，而是几十次，直到 Telegram 停止重新尝试。
```

File: site/docs/zh/guide/deployment-types.md [hunk 22]
```diff
@@ -1,3 +1,387 @@
...
+你可能会发现你的 bot 开始给用户发送垃圾邮件，因为它试图处理所有这些更新(实际上每次都是一样的)。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:325]:
```suggestion
你可能会发现你的 bot 开始给用户发送垃圾信息，因为它试图处理所有这些 update（实际上每次都是一样的）。
```

File: site/docs/zh/guide/deployment-types.md [hunk 23]
```diff
@@ -1,3 +1,387 @@
...
+
+#### 为什么提前终止一个 Webhook 请求也是危险的
+
+你可以配置 `webhookCallback` 在超时后不抛出错误，而是提前结束 webhook 请求，即使你的中间件仍然在运行。
+你可以通过传递 `return` 作为第三个参数到 `webhookCallback`，而不是默认使用 `throw` 来实现。
+然而，尽管这种行为有一些有效的用例，但这种解决方案通常会导致比它解决的问题更多的问题。
+
+请记住，一旦你响应了一个 webhook 请求，Telegram 将发送该聊天的下一个更新。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:333]:
```suggestion
请记住，一旦你响应了一个 webhook 请求，Telegram 将发送该聊天的下一个 update。
```

File: site/docs/zh/guide/deployment-types.md [hunk 24]
```diff
@@ -1,3 +1,387 @@
...
+但是，由于旧的更新仍在处理中，先前顺序处理的两个更新突然并行处理。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:334]:
```suggestion
但是，由于旧的 update 仍在处理中，先前顺序处理的两个 update 突然并行处理。
```

File: site/docs/zh/guide/deployment-types.md [hunk 25]
```diff
@@ -1,3 +1,387 @@
...
+
+#### 为什么 `return` 比 `throw` 更糟糕
+
+你可能想知道为什么 `webhookCallback` 的默认操作是 throw 错误，而不是成功结束请求。
+这个设计的选择是基于以下原因。
+
+竞态效应极少或偶尔发生。
+解决这个问题的方法是首先确保不会遇到超时。
+不过，如果你这样做了，你将知道这正在发生，这样你就可以调查并解决问题！
+由于这个原因，您希望错误可以被记录在日志中。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:380]:
```suggestion
由于这个原因，你希望错误可以被记录在日志中。
```

File: site/docs/zh/guide/deployment-types.md [hunk 26]
```diff
@@ -1,3 +1,387 @@
...
+将超时处理程序设置为“ return”，从而抑制超时并假装什么都没有发生，这与有用的行为恰恰相反。
+
+如果你这样做，在某种意义上，你是在使用 Telegram 的 webhook 传递中的更新队列作为你的任务队列。
+由于上述所有原因，这是一个坏主意。
+仅仅因为 grammY 可以抑制可能导致你丢失数据的错误，但这并不意味着你应该告诉它。
+在您的中间件需要花费太多时间才能完成的情况下，不应该使用此配置设置。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:386]:
```suggestion
在你的中间件需要花费太多时间才能完成的情况下，不应该使用此配置设置。
```

File: site/docs/zh/guide/deployment-types.md [hunk 27]
```diff
@@ -1,3 +1,387 @@
...
+你可能想知道为什么 `webhookCallback` 的默认操作是 throw 错误，而不是成功结束请求。
+这个设计的选择是基于以下原因。
+
+竞态效应极少或偶尔发生。
+解决这个问题的方法是首先确保不会遇到超时。
+不过，如果你这样做了，你将知道这正在发生，这样你就可以调查并解决问题！
+由于这个原因，您希望错误可以被记录在日志中。
+将超时处理程序设置为“ return”，从而抑制超时并假装什么都没有发生，这与有用的行为恰恰相反。
+
+如果你这样做，在某种意义上，你是在使用 Telegram 的 webhook 传递中的更新队列作为你的任务队列。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:383]:
```suggestion
如果你这样做，在某种意义上，你是在使用 Telegram 的 webhook 传递中的 update 队列作为你的任务队列。
```

File: site/docs/zh/guide/deployment-types.md [hunk 28]
```diff
@@ -1,3 +1,387 @@
...
+漂亮。
+这就是当你运行 `bot.start ()` 时，默认情况下 grammY 会做的事情。
+
+## Webhooks 是如何工作的？
+
+_在这次可怕的经历之后(一整晚都没有冰淇淋!)，你不再问任何人关于冰淇淋的事。
+但如果冰淇淋可以送到你面前，那不是很酷吗_
+
+建立一个 webhook 意味着你将为 Telegram 提供一个可以从公共互联网上访问的 URL。
+无论何时，只要有新的信息发送到你的 bot，Telegram (而不是你!) 将采取主动，并发送一个请求与更新对象到您的服务器。
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:112]:
```suggestion
无论何时，只要有新的信息发送到你的 bot，Telegram（而不是你！）将采取主动，并发送一个请求与 update 对象到你的服务器。
```

File: site/docs/zh/guide/deployment-types.md [hunk 29]
```diff
@@ -1,3 +1,387 @@
...
+
+</CodeGroupItem>
+</CodeGroup>
+
+如果你考虑在 VPS 上使用 webhooks 运行你的 bot，请确保你阅读了 Telegram 团队写的 [Marvin's Marvellous Guide to All Things Webhook](https://core.telegram.org/bots/webhooks)。
+
+### Webhook Reply
+
+当收到一个 webhook 请求时，您的 bot 可以在响应中调用一个方法。
+这样做的一个好处是，每次更新都可以节省一个 HTTP 请求的 bot。然而，使用这种方法也有一些缺点:
```
Comment (WingLim) [inline site/docs/zh/guide/deployment-types.md:239]:
```suggestion
这样做的一个好处是，每个 update 都可以节省一次 HTTP 请求。然而，使用这种方法也有一些缺点:
```

**Review by WingLim — CHANGES_REQUESTED** [review]
`更新` --> `update`
`您` --> `你`
和一些非中文符号~

**Comment by linbuxiao** [comment]
> Maybe have a look at https://github.com/nvm-sh/nvm. It's the only way I know that makes handing different versions of Node and npm feasible. It gets super simple to select the right version of npm such that this doesn't happen.

Thank you! Switching versions helped me solve the problem.

**Review by WingLim — APPROVED** [review]
Great job!

**Pull request merged**